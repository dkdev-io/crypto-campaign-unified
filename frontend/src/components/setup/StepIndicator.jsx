import React from 'react';

const StepIndicator = ({ currentStep, totalSteps }) => {
  const steps = [
    'Campaign Info',
    'Committee Search',
    'Bank Connection',
    'Form Customization',
    'Style Confirmation',
    'Launch & QR Code',
    'Terms Agreement',
  ];

  return (
    <div className="step-indicator">
      <h1>Campaign Setup</h1>
      <p>
        Step {currentStep} of {totalSteps}: {steps[currentStep - 1]}
      </p>
    </div>
  );
};

export default StepIndicator;
