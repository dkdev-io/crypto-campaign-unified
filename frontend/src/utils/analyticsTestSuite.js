/**
 * Analytics Test Suite
 * Comprehensive testing utilities for the campaign analytics system
 */

import CampaignAnalytics from './campaignAnalytics';

class AnalyticsTestSuite {
  constructor() {
    this.testResults = [];
    this.analytics = null;
    this.mockConsole = {
      log: (...args) => this.testResults.push({ type: 'log', message: args.join(' ') }),
      error: (...args) => this.testResults.push({ type: 'error', message: args.join(' ') }),
      warn: (...args) => this.testResults.push({ type: 'warn', message: args.join(' ') }),
    };
  }

  /**
   * Initialize test environment
   */
  async setup() {
    // Mock localStorage
    global.localStorage = {
      data: {},
      getItem: (key) => global.localStorage.data[key] || null,
      setItem: (key, value) => (global.localStorage.data[key] = value),
      removeItem: (key) => delete global.localStorage.data[key],
      clear: () => (global.localStorage.data = {}),
    };

    // Mock sessionStorage
    global.sessionStorage = {
      data: {},
      getItem: (key) => global.sessionStorage.data[key] || null,
      setItem: (key, value) => (global.sessionStorage.data[key] = value),
      removeItem: (key) => delete global.sessionStorage.data[key],
      clear: () => (global.sessionStorage.data = {}),
    };

    // Mock document and window
    global.document = {
      readyState: 'complete',
      referrer: 'https://google.com/search',
      title: 'Test Campaign Page',
      addEventListener: () => {},
      getElementById: () => null,
      createElement: () => ({
        style: {},
        innerHTML: '',
        onclick: null,
        remove: () => {},
        addEventListener: () => {},
      }),
      body: { appendChild: () => {} },
    };

    global.window = {
      location: {
        href: 'https://example.com/campaign/test-campaign?utm_source=google&utm_medium=cpc',
        pathname: '/campaign/test-campaign',
        search: '?utm_source=google&utm_medium=cpc',
      },
      innerWidth: 1920,
      innerHeight: 1080,
      pageYOffset: 0,
      addEventListener: () => {},
      removeEventListener: () => {},
    };

    global.navigator = {
      userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
      doNotTrack: '0',
      sendBeacon: () => true,
    };

    global.screen = {
      width: 2560,
      height: 1440,
    };

    // Mock fetch for API calls
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ success: true }),
      })
    );

    console.log('Test environment initialized');
  }

  /**
   * Run all tests
   */
  async runAllTests() {
    await this.setup();

    console.log('🧪 Starting Analytics Test Suite...\n');

    const tests = [
      'testVisitorIdGeneration',
      'testSessionManagement',
      'testConsentManagement',
      'testEventTracking',
      'testTrafficSourceDetection',
      'testFormInteractionTracking',
      'testConversionTracking',
      'testPrivacyCompliance',
      'testErrorHandling',
      'testPerformance',
    ];

    for (const testName of tests) {
      try {
        await this[testName]();
        console.log(`✅ ${testName} - PASSED`);
      } catch (error) {
        console.error(`❌ ${testName} - FAILED:`, error.message);
      }
    }

    this.printSummary();
  }

  /**
   * Test visitor ID generation and persistence
   */
  async testVisitorIdGeneration() {
    // Clear storage
    localStorage.clear();

    // Create analytics instance
    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics.initializeVisitor();

    const visitorId1 = analytics.visitorId;
    this.assert(visitorId1, 'Visitor ID should be generated');
    this.assert(visitorId1.startsWith('visitor_'), 'Visitor ID should have correct prefix');

    // Test persistence
    const analytics2 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });
    await analytics2.initializeVisitor();

    const visitorId2 = analytics2.visitorId;
    this.assert(visitorId1 === visitorId2, 'Visitor ID should persist across instances');
  }

  /**
   * Test session management
   */
  async testSessionManagement() {
    sessionStorage.clear();

    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
      sessionTimeout: 1, // 1 minute for testing
    });

    await analytics.initializeSession();

    const sessionId1 = analytics.sessionId;
    this.assert(sessionId1, 'Session ID should be generated');
    this.assert(sessionId1.startsWith('session_'), 'Session ID should have correct prefix');
    this.assert(analytics.sessionStart instanceof Date, 'Session start should be a Date');

    // Test session continuation
    const analytics2 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
      sessionTimeout: 1,
    });
    await analytics2.initializeSession();

    this.assert(analytics2.sessionId === sessionId1, 'Session should continue if within timeout');
  }

  /**
   * Test consent management
   */
  async testConsentManagement() {
    localStorage.clear();

    // Test optional consent (default allow)
    const analytics1 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'optional',
    });
    await analytics1.checkConsent();
    this.assert(analytics1.consentGiven !== false, 'Optional consent should default to allow');

    // Test required consent
    const analytics2 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'required',
    });

    // Mock user denying consent
    analytics2.requestConsent = () => Promise.resolve(false);
    await analytics2.checkConsent();
    this.assert(analytics2.consentGiven === false, 'Required consent should respect user choice');

    // Test disabled consent
    const analytics3 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });
    await analytics3.checkConsent();
    this.assert(analytics3.consentGiven === true, 'Disabled consent should always allow');
  }

  /**
   * Test event tracking
   */
  async testEventTracking() {
    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics.init();

    // Clear events
    analytics.events = [];

    // Track custom event
    analytics.trackEvent('test_event', { key: 'value' });

    this.assert(analytics.events.length === 1, 'Event should be tracked');

    const event = analytics.events[0];
    this.assert(event.event_type === 'test_event', 'Event type should match');
    this.assert(event.event_data.key === 'value', 'Event data should be preserved');
    this.assert(event.visitor_id === analytics.visitorId, 'Event should include visitor ID');
    this.assert(event.session_id === analytics.sessionId, 'Event should include session ID');
  }

  /**
   * Test traffic source detection
   */
  async testTrafficSourceDetection() {
    // Test UTM parameters
    window.location.search = '?utm_source=facebook&utm_medium=social&utm_campaign=spring2024';

    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    const trafficSource = analytics.getTrafficSource();

    this.assert(trafficSource.traffic_source === 'facebook', 'Should detect UTM source');
    this.assert(trafficSource.traffic_medium === 'social', 'Should detect UTM medium');
    this.assert(trafficSource.traffic_campaign === 'spring2024', 'Should detect UTM campaign');

    // Test referrer detection
    window.location.search = '';
    document.referrer = 'https://twitter.com/somepost';

    const analytics2 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    const trafficSource2 = analytics2.getTrafficSource();
    this.assert(trafficSource2.traffic_source === 'twitter', 'Should detect Twitter referrer');
    this.assert(
      trafficSource2.traffic_medium === 'social',
      'Should classify social media correctly'
    );

    // Test direct traffic
    document.referrer = '';

    const analytics3 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    const trafficSource3 = analytics3.getTrafficSource();
    this.assert(trafficSource3.traffic_source === 'direct', 'Should detect direct traffic');
  }

  /**
   * Test form interaction tracking
   */
  async testFormInteractionTracking() {
    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics.init();
    analytics.events = [];

    // Mock form elements
    const mockInput = {
      tagName: 'INPUT',
      id: 'contribution-amount',
      type: 'number',
      value: '1.5',
      name: 'amount',
    };

    const mockForm = {
      tagName: 'FORM',
      id: 'contribution-form',
    };

    // Simulate focus event
    const focusEvent = { target: mockInput };
    document.addEventListener = (event, handler) => {
      if (event === 'focus') {
        handler(focusEvent);
      }
    };

    analytics.setupFormTracking();

    // Check if form interaction was tracked
    this.assert(
      analytics.formInteractions.has('contribution-amount'),
      'Form field should be tracked'
    );
  }

  /**
   * Test conversion tracking
   */
  async testConversionTracking() {
    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics.init();
    analytics.events = [];

    const conversionData = {
      amount: 2.5,
      transaction_hash: '0x123abc',
      currency: 'ETH',
    };

    analytics.trackConversion(conversionData);

    this.assert(analytics.events.length === 1, 'Conversion event should be tracked');

    const event = analytics.events[0];
    this.assert(event.event_type === 'conversion', 'Event type should be conversion');
    this.assert(event.event_data.amount === 2.5, 'Amount should be preserved');
    this.assert(
      event.event_data.transaction_hash === '0x123abc',
      'Transaction hash should be preserved'
    );
  }

  /**
   * Test privacy compliance features
   */
  async testPrivacyCompliance() {
    // Test Do Not Track respect
    navigator.doNotTrack = '1';

    const analytics1 = new CampaignAnalytics({
      debug: false,
      respectDNT: true,
      cookieConsent: 'disabled',
    });

    await analytics1.init();
    this.assert(analytics1.consentGiven !== true, 'Should respect Do Not Track');

    // Test data clearing
    navigator.doNotTrack = '0';

    const analytics2 = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics2.init();
    analytics2.trackEvent('test_event');

    this.assert(analytics2.events.length > 0, 'Should have events before clearing');

    analytics2.clearAllData();

    this.assert(analytics2.events.length === 0, 'Events should be cleared');
    this.assert(analytics2.visitorId === null, 'Visitor ID should be cleared');
    this.assert(localStorage.getItem('visitor_id') === null, 'Storage should be cleared');
  }

  /**
   * Test error handling
   */
  async testErrorHandling() {
    // Mock fetch failure
    const originalFetch = global.fetch;
    global.fetch = jest.fn(() => Promise.reject(new Error('Network error')));

    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
    });

    await analytics.init();
    analytics.trackEvent('test_event');

    // Should not throw error
    try {
      await analytics.flushEvents();
      console.log('Error handling test passed - no exception thrown');
    } catch (error) {
      throw new Error('Should handle network errors gracefully');
    }

    // Restore fetch
    global.fetch = originalFetch;
  }

  /**
   * Test performance characteristics
   */
  async testPerformance() {
    const analytics = new CampaignAnalytics({
      debug: false,
      cookieConsent: 'disabled',
      batchSize: 5,
    });

    await analytics.init();
    analytics.events = [];

    // Test batch processing
    const startTime = Date.now();

    for (let i = 0; i < 10; i++) {
      analytics.trackEvent(`test_event_${i}`, { index: i });
    }

    const endTime = Date.now();
    const processingTime = endTime - startTime;

    this.assert(processingTime < 100, `Event tracking should be fast (${processingTime}ms)`);
    this.assert(analytics.events.length <= 5, 'Should trigger batch send at batch size');

    // Test ID generation performance
    const idStartTime = Date.now();
    const ids = [];

    for (let i = 0; i < 1000; i++) {
      ids.push(analytics.generateId('test'));
    }

    const idEndTime = Date.now();
    const idGenerationTime = idEndTime - idStartTime;

    this.assert(idGenerationTime < 50, `ID generation should be fast (${idGenerationTime}ms)`);
    this.assert(new Set(ids).size === 1000, 'All generated IDs should be unique');
  }

  /**
   * Test utility - assertion helper
   */
  assert(condition, message) {
    if (!condition) {
      throw new Error(message);
    }
  }

  /**
   * Print test summary
   */
  printSummary() {
    const logs = this.testResults.filter((r) => r.type === 'log');
    const errors = this.testResults.filter((r) => r.type === 'error');

    console.log('\n📊 Test Summary:');
    console.log(`   Logs: ${logs.length}`);
    console.log(`   Errors: ${errors.length}`);

    if (errors.length > 0) {
      console.log('\n❌ Errors:');
      errors.forEach((error) => console.log(`   - ${error.message}`));
    }

    console.log('\n🎉 Analytics Test Suite Complete!');
  }
}

/**
 * Manual testing utilities for browser environment
 */
export const manualTests = {
  /**
   * Test analytics initialization in browser
   */
  testInitialization: async () => {
    console.log('🧪 Testing analytics initialization...');

    const analytics = new CampaignAnalytics({
      debug: true,
      cookieConsent: 'disabled',
    });

    await analytics.init();

    console.log('✅ Analytics initialized');
    console.log('Visitor ID:', analytics.visitorId);
    console.log('Session ID:', analytics.sessionId);
    console.log('Status:', analytics.getTrackingStatus());
  },

  /**
   * Test event tracking
   */
  testEventTracking: () => {
    console.log('🧪 Testing event tracking...');

    if (window.trackEvent) {
      window.trackEvent('manual_test_event', {
        test: true,
        timestamp: Date.now(),
      });

      console.log('✅ Event tracked via global function');
    } else {
      console.error('❌ Global trackEvent function not available');
    }
  },

  /**
   * Test form interactions
   */
  testFormInteractions: () => {
    console.log('🧪 Testing form interactions...');

    // Create test form elements
    const form = document.createElement('form');
    form.id = 'test-form';

    const input = document.createElement('input');
    input.type = 'text';
    input.id = 'test-input';
    input.placeholder = 'Test input field';

    const button = document.createElement('button');
    button.type = 'submit';
    button.textContent = 'Test Submit';

    form.appendChild(input);
    form.appendChild(button);
    document.body.appendChild(form);

    // Test focus event
    input.focus();

    // Test form submission
    form.addEventListener('submit', (e) => {
      e.preventDefault();
      console.log('✅ Form interaction events should be tracked');
    });

    setTimeout(() => {
      form.dispatchEvent(new Event('submit'));
      document.body.removeChild(form);
    }, 1000);
  },

  /**
   * Test conversion tracking
   */
  testConversionTracking: () => {
    console.log('🧪 Testing conversion tracking...');

    if (window.trackConversion) {
      window.trackConversion({
        amount: 1.5,
        transaction_hash: '0xtest123',
        currency: 'ETH',
        test: true,
      });

      console.log('✅ Conversion tracked via global function');
    } else {
      console.error('❌ Global trackConversion function not available');
    }
  },

  /**
   * Test privacy controls
   */
  testPrivacyControls: () => {
    console.log('🧪 Testing privacy controls...');

    if (window.setAnalyticsConsent) {
      // Test disabling
      window.setAnalyticsConsent(false);
      console.log('Analytics disabled');

      // Test enabling
      setTimeout(() => {
        window.setAnalyticsConsent(true);
        console.log('Analytics enabled');
        console.log('✅ Privacy controls working');
      }, 2000);
    } else {
      console.error('❌ Global setAnalyticsConsent function not available');
    }
  },

  /**
   * Run all manual tests
   */
  runAll: async () => {
    await manualTests.testInitialization();
    setTimeout(() => manualTests.testEventTracking(), 1000);
    setTimeout(() => manualTests.testFormInteractions(), 2000);
    setTimeout(() => manualTests.testConversionTracking(), 4000);
    setTimeout(() => manualTests.testPrivacyControls(), 5000);
  },
};

// Export test suite
export default AnalyticsTestSuite;

// Auto-run tests in Node.js environment
if (typeof window === 'undefined' && typeof process !== 'undefined') {
  const testSuite = new AnalyticsTestSuite();
  testSuite.runAllTests().catch(console.error);
}
