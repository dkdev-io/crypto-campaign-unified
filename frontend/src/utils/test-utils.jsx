import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { vi } from 'vitest';

// Test data factories
export const createMockCampaign = (overrides = {}) => ({
  id: 'test-campaign-1',
  campaign_name: 'Test Campaign',
  email: 'test@example.com',
  candidate_name: 'John Doe',
  theme_color: '#2a2a72',
  suggested_amounts: [25, 50, 100, 250],
  max_donation_limit: 3300,
  website: 'https://test.com',
  wallet_address: '0x123...',
  created_at: '2024-01-01T00:00:00.000Z',
  ...overrides
});

export const createMockDonorData = (overrides = {}) => ({
  fullName: 'Jane Smith',
  email: 'jane@example.com',
  phone: '555-123-4567',
  street: '123 Main St',
  city: 'Anytown',
  state: 'CA',
  zip: '12345',
  employer: 'Test Corp',
  occupation: 'Engineer',
  amount: '100',
  ...overrides
});

export const createMockWalletInfo = (overrides = {}) => ({
  isConnected: true,
  account: '0x742d35Cc6C4C5e5D0E9dF5F9B6A5e8a2c1B9D8e7',
  balance: '1.5678',
  network: {
    name: 'Ethereum Mainnet',
    chainId: '0x1'
  },
  contributorInfo: {
    isKYCVerified: true,
    cumulativeAmount: 0.5,
    remainingCapacity: 1.0
  },
  ...overrides
});

export const createMockFormSubmission = (overrides = {}) => ({
  id: 'submission-1',
  campaign_id: 'test-campaign-1',
  donor_full_name: 'Jane Smith',
  donor_email: 'jane@example.com',
  amount_usd: 100,
  transaction_hash: '0xabc123...',
  status: 'confirmed',
  created_at: '2024-01-01T12:00:00.000Z',
  ...overrides
});

// Custom render function with providers
export const render = (ui, {
  route = '/',
  ...renderOptions
} = {}) => {
  // Set initial URL for router
  window.history.pushState({}, 'Test page', route);
  
  const Wrapper = ({ children }) => (
    <BrowserRouter>
      {children}
    </BrowserRouter>
  );

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// Utility to wait for async operations
export const waitForAsync = () => new Promise(resolve => setTimeout(resolve, 0));

// Mock event creators
export const createMockEvent = (overrides = {}) => ({
  preventDefault: vi.fn(),
  stopPropagation: vi.fn(),
  target: {
    value: 'test-value',
    name: 'test-input',
    checked: false,
    ...overrides.target
  },
  currentTarget: {
    value: 'test-value',
    ...overrides.currentTarget
  },
  ...overrides
});

export const createChangeEvent = (value, name = 'test-input') => 
  createMockEvent({
    target: { value, name },
    currentTarget: { value }
  });

export const createClickEvent = (overrides = {}) =>
  createMockEvent({
    type: 'click',
    ...overrides
  });

export const createSubmitEvent = (overrides = {}) =>
  createMockEvent({
    type: 'submit',
    ...overrides
  });

// Form validation helpers
export const fillFormFields = async (user, formData) => {
  const fields = Object.entries(formData);
  
  for (const [fieldName, value] of fields) {
    const input = document.querySelector(`input[name="${fieldName}"], select[name="${fieldName}"], textarea[name="${fieldName}"]`);
    if (input) {
      if (input.type === 'checkbox' || input.type === 'radio') {
        if (value) {
          await user.click(input);
        }
      } else {
        await user.type(input, String(value));
      }
    }
  }
};

// Assertion helpers
export const expectFormField = (container, fieldName, expectedValue = null) => {
  const field = container.querySelector(`[name="${fieldName}"]`);
  expect(field).toBeInTheDocument();
  
  if (expectedValue !== null) {
    if (field.type === 'checkbox' || field.type === 'radio') {
      expect(field.checked).toBe(Boolean(expectedValue));
    } else {
      expect(field.value).toBe(String(expectedValue));
    }
  }
  
  return field;
};

export const expectErrorMessage = (container, message) => {
  const errorElement = container.querySelector('[style*="background: #f8d7da"], [style*="color: #721c24"], .error-message');
  expect(errorElement).toBeInTheDocument();
  if (message) {
    expect(errorElement).toHaveTextContent(message);
  }
};

export const expectSuccessMessage = (container, message) => {
  const successElement = container.querySelector('[style*="background: #d4edda"], [style*="color: #155724"], .success-message');
  expect(successElement).toBeInTheDocument();
  if (message) {
    expect(successElement).toHaveTextContent(message);
  }
};

// Mock HTTP responses
export const mockApiResponse = (data, options = {}) => {
  const {
    status = 200,
    ok = true,
    delay = 0
  } = options;
  
  return new Promise(resolve => {
    setTimeout(() => {
      resolve({
        ok,
        status,
        statusText: ok ? 'OK' : 'Error',
        json: () => Promise.resolve(data),
        text: () => Promise.resolve(JSON.stringify(data))
      });
    }, delay);
  });
};

// Component test helpers
export const getByDataTestId = (container, testId) => {
  return container.querySelector(`[data-testid="${testId}"]`);
};

export const getAllByDataTestId = (container, testId) => {
  return container.querySelectorAll(`[data-testid="${testId}"]`);
};

// Form interaction patterns
export const submitForm = async (user, form) => {
  const submitButton = form.querySelector('button[type="submit"], input[type="submit"]');
  if (submitButton) {
    await user.click(submitButton);
  } else {
    // Submit via form if no submit button found
    await user.submit(form);
  }
};

export const selectSuggestedAmount = async (user, container, amount) => {
  const amountButton = container.querySelector(`button[type="button"]:has-text("$${amount}")`);
  if (amountButton) {
    await user.click(amountButton);
  }
  return amountButton;
};

// Wallet interaction helpers
export const connectWallet = async (user, container) => {
  const connectButton = container.querySelector('button:has-text("Connect"), button:has-text("MetaMask")');
  if (connectButton) {
    await user.click(connectButton);
  }
  return connectButton;
};

export const disconnectWallet = async (user, container) => {
  const disconnectButton = container.querySelector('button:has-text("Disconnect")');
  if (disconnectButton) {
    await user.click(disconnectButton);
  }
  return disconnectButton;
};

// Time utilities for testing
export const mockDateNow = (timestamp) => {
  const dateSpy = vi.spyOn(Date, 'now');
  dateSpy.mockReturnValue(timestamp);
  return dateSpy;
};

export const mockDate = (dateString) => {
  const mockDate = new Date(dateString);
  const dateSpy = vi.spyOn(global, 'Date');
  dateSpy.mockImplementation(() => mockDate);
  return dateSpy;
};

// Export everything for easy importing
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { vi } from 'vitest';