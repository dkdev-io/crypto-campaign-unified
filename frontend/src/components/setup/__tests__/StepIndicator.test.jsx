import { describe, it, expect, beforeEach, vi } from 'vitest';
import { render, screen } from '../../../utils/test-utils';
import StepIndicator from '../StepIndicator';

describe('StepIndicator', () => {
  describe('Component Rendering', () => {
    it('renders step indicator with current and total steps', () => {
      render(<StepIndicator currentStep={1} totalSteps={5} />);

      expect(screen.getByText('Step 1 of 5')).toBeInTheDocument();
    });

    it('renders progress bar', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={4} />);

      expect(container.querySelector('.step-indicator')).toBeInTheDocument();
      expect(container.querySelector('.progress-bar')).toBeInTheDocument();
    });

    it('displays individual step circles', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);

      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(5);
    });
  });

  describe('Step States', () => {
    it('marks completed steps correctly', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);

      const stepCircles = container.querySelectorAll('.step-circle');

      // Steps 1 and 2 should be completed
      expect(stepCircles[0]).toHaveClass('completed');
      expect(stepCircles[1]).toHaveClass('completed');

      // Step 3 should be active (current)
      expect(stepCircles[2]).toHaveClass('active');

      // Steps 4 and 5 should be inactive
      expect(stepCircles[3]).not.toHaveClass('completed');
      expect(stepCircles[3]).not.toHaveClass('active');
      expect(stepCircles[4]).not.toHaveClass('completed');
      expect(stepCircles[4]).not.toHaveClass('active');
    });

    it('marks current step as active', () => {
      const { container } = render(<StepIndicator currentStep={4} totalSteps={6} />);

      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles[3]).toHaveClass('active'); // 4th step (index 3)
    });

    it('shows correct step numbers', () => {
      render(<StepIndicator currentStep={2} totalSteps={4} />);

      expect(screen.getByText('1')).toBeInTheDocument();
      expect(screen.getByText('2')).toBeInTheDocument();
      expect(screen.getByText('3')).toBeInTheDocument();
      expect(screen.getByText('4')).toBeInTheDocument();
    });
  });

  describe('Progress Bar', () => {
    it('calculates correct progress percentage', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);

      const progressFill = container.querySelector('.progress-fill');
      // Progress should be (currentStep - 1) / (totalSteps - 1) * 100 = 2/4 * 100 = 50%
      expect(progressFill).toHaveStyle({ width: '50%' });
    });

    it('shows 0% progress on first step', () => {
      const { container } = render(<StepIndicator currentStep={1} totalSteps={5} />);

      const progressFill = container.querySelector('.progress-fill');
      expect(progressFill).toHaveStyle({ width: '0%' });
    });

    it('shows 100% progress on last step', () => {
      const { container } = render(<StepIndicator currentStep={5} totalSteps={5} />);

      const progressFill = container.querySelector('.progress-fill');
      expect(progressFill).toHaveStyle({ width: '100%' });
    });
  });

  describe('Different Step Configurations', () => {
    it('handles 3-step process', () => {
      render(<StepIndicator currentStep={2} totalSteps={3} />);

      expect(screen.getByText('Step 2 of 3')).toBeInTheDocument();

      const { container } = render(<StepIndicator currentStep={2} totalSteps={3} />);
      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(3);
    });

    it('handles 7-step process', () => {
      render(<StepIndicator currentStep={4} totalSteps={7} />);

      expect(screen.getByText('Step 4 of 7')).toBeInTheDocument();

      const { container } = render(<StepIndicator currentStep={4} totalSteps={7} />);
      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(7);
    });

    it('handles single step process', () => {
      render(<StepIndicator currentStep={1} totalSteps={1} />);

      expect(screen.getByText('Step 1 of 1')).toBeInTheDocument();

      const { container } = render(<StepIndicator currentStep={1} totalSteps={1} />);
      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(1);
      expect(stepCircles[0]).toHaveClass('active');
    });
  });

  describe('CSS Classes and Styling', () => {
    it('applies correct base CSS classes', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={4} />);

      expect(container.querySelector('.step-indicator')).toBeInTheDocument();
      expect(container.querySelector('.step-text')).toBeInTheDocument();
      expect(container.querySelector('.steps-container')).toBeInTheDocument();
    });

    it('applies responsive styling classes', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={6} />);

      const stepsContainer = container.querySelector('.steps-container');
      expect(stepsContainer).toHaveClass('steps-container');
    });

    it('applies correct colors to step states', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);

      const stepCircles = container.querySelectorAll('.step-circle');

      // Completed steps should have completed styling
      expect(stepCircles[0]).toHaveClass('completed');
      expect(stepCircles[1]).toHaveClass('completed');

      // Active step should have active styling
      expect(stepCircles[2]).toHaveClass('active');
    });
  });

  describe('Edge Cases', () => {
    it('handles step 0 gracefully', () => {
      render(<StepIndicator currentStep={0} totalSteps={5} />);

      expect(screen.getByText('Step 0 of 5')).toBeInTheDocument();
    });

    it('handles current step greater than total steps', () => {
      render(<StepIndicator currentStep={6} totalSteps={5} />);

      expect(screen.getByText('Step 6 of 5')).toBeInTheDocument();

      const { container } = render(<StepIndicator currentStep={6} totalSteps={5} />);
      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(5); // Should still render 5 circles
    });

    it('handles negative step numbers', () => {
      render(<StepIndicator currentStep={-1} totalSteps={3} />);

      expect(screen.getByText('Step -1 of 3')).toBeInTheDocument();
    });

    it('handles zero total steps', () => {
      render(<StepIndicator currentStep={1} totalSteps={0} />);

      expect(screen.getByText('Step 1 of 0')).toBeInTheDocument();

      const { container } = render(<StepIndicator currentStep={1} totalSteps={0} />);
      const stepCircles = container.querySelectorAll('.step-circle');
      expect(stepCircles).toHaveLength(0);
    });
  });

  describe('Step Labels', () => {
    it('generates correct step labels for 5-step process', () => {
      render(<StepIndicator currentStep={3} totalSteps={5} />);

      const stepLabels = ['Setup', 'Committee', 'Banking', 'Terms', 'Launch'];

      stepLabels.forEach((label, index) => {
        expect(screen.getByText(label)).toBeInTheDocument();
      });
    });

    it('shows generic labels for non-standard step counts', () => {
      render(<StepIndicator currentStep={2} totalSteps={3} />);

      expect(screen.getByText('Step 1')).toBeInTheDocument();
      expect(screen.getByText('Step 2')).toBeInTheDocument();
      expect(screen.getByText('Step 3')).toBeInTheDocument();
    });

    it('truncates labels on small screens', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={5} />);

      const stepLabels = container.querySelectorAll('.step-label');
      stepLabels.forEach((label) => {
        expect(label).toHaveClass('step-label');
      });
    });
  });

  describe('Accessibility', () => {
    it('has proper ARIA labels', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);

      const stepIndicator = container.querySelector('.step-indicator');
      expect(stepIndicator).toHaveAttribute('aria-label', 'Step 3 of 5');
    });

    it('marks current step with aria-current', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={4} />);

      const activeStep = container.querySelector('.step-circle.active');
      expect(activeStep).toHaveAttribute('aria-current', 'step');
    });

    it('provides accessible text for screen readers', () => {
      render(<StepIndicator currentStep={3} totalSteps={5} />);

      expect(screen.getByText('Step 3 of 5')).toBeInTheDocument();
    });

    it('has proper role attributes', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={4} />);

      const stepsContainer = container.querySelector('.steps-container');
      expect(stepsContainer).toHaveAttribute('role', 'progressbar');
    });
  });

  describe('Component Snapshots', () => {
    it('matches snapshot for beginning of process', () => {
      const { container } = render(<StepIndicator currentStep={1} totalSteps={5} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for middle of process', () => {
      const { container } = render(<StepIndicator currentStep={3} totalSteps={5} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for end of process', () => {
      const { container } = render(<StepIndicator currentStep={5} totalSteps={5} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for different step count', () => {
      const { container } = render(<StepIndicator currentStep={2} totalSteps={3} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
