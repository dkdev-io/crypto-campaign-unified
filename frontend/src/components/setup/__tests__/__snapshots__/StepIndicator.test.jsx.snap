// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StepIndicator > Component Snapshots > matches snapshot for beginning of process 1`] = `
<div
  class="step-indicator"
>
  <h1>
    Campaign Setup
  </h1>
  <p>
    Step 
    1
     of 
    5
    : 
    Signup
  </p>
</div>
`;

exports[`StepIndicator > Component Snapshots > matches snapshot for different step count 1`] = `
<div
  class="step-indicator"
>
  <h1>
    Campaign Setup
  </h1>
  <p>
    Step 
    2
     of 
    3
    : 
    Campaign Info
  </p>
</div>
`;

exports[`StepIndicator > Component Snapshots > matches snapshot for end of process 1`] = `
<div
  class="step-indicator"
>
  <h1>
    Campaign Setup
  </h1>
  <p>
    Step 
    5
     of 
    5
    : 
    Embed Options
  </p>
</div>
`;

exports[`StepIndicator > Component Snapshots > matches snapshot for middle of process 1`] = `
<div
  class="step-indicator"
>
  <h1>
    Campaign Setup
  </h1>
  <p>
    Step 
    3
     of 
    5
    : 
    Compliance
  </p>
</div>
`;
